### ISSUE TITLE
[Draft][API Design] Query hints, options.

### ISSUE THREAD

**Issue by sdanyliv** [linq2db/linq2db#2452]
Opened this issue for discussion about API design. Feel free to correct spelling and comment missed parts.
It is completely draft and can be changed frequently depending on found ambiguous scenarios.

##### Sample query
```cs
var query = 
   from t1 in table1
   join t2 in table2 on t1.Id equals t2.Id
   select new { t1, t2 }
```
---
### Oracle
https://docs.oracle.com/cd/B13789_01/server.101/b10752/hintsref.htm#30459

`namespace LinqToDB.DataProvider.Oracle;`
##### Any query hint
```cs
query = query.StatementHint(q => $"LEADING({q.t1} {q.t2}) USE_NL({q.t1}) INDEX({q.t1} emp_emp_id_pk) USE_MERGE({q.t1}) FULL({q.t1})")
```
##### Flashback query
https://oracle-base.com/articles/10g/flashback-query-10g
```cs
var query = 
   from t1 in table1.AsOfScn(scnNumber)
   join t2 in table2.AsOfTimestamp(timestamp) on t1.Id equals t2.Id
   select new { t1, t2 }
```
`AsOfScn/AsOfTimestamp` applied to query should apply `AS OF` table hint to all tables in query. So the following queries are equivalent:
```cs
var query = query.AsOfScn(scnNumber)
```
```cs
var query = 
   from t1 in table1.AsOfScn(scnNumber)
   join t2 in table2.AsOfScn(scnNumber) on t1.Id equals t2.Id
   select new { t1, t2 }
```
---
### PostgreSQL

Looks like for PostgresSQL we can only define locks after select statement:
https://www.postgresql.org/docs/current/sql-select.html
Consider to add `AfterHint` which should add text exactly after whole query:
```cs
var query = query.AfterHint(q => $"FOR KEY UPDATE OF {q.t1}");
```
---
### MySql

Looks like the same hints as for Oracle
https://dev.mysql.com/doc/refman/8.0/en/optimizer-hints.html#optimizer-hints-syntax

---
### Informix

Looks like the same hints as for Oracle but divided by comma
https://www.ibm.com/developerworks/data/zones/informix/library/techarticle/0502fan/0502fan.html

---
### SqLite

Currently found only INDEXED BY injected before table name
https://www.tutlane.com/tutorial/sqlite/sqlite-indexed-by

Consider to add `IndexedBy` extension for `ITable<>`
```cs
var query = 
   from t1 in table1.IndexedBy("index_name")
   join t2 in table2 on t1.Id equals t2.Id
   select new { t1, t2 }
```

---
### SqlServer

https://docs.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql?view=sql-server-ver15

**Comment by sdanyliv** [comment]
@filippobottega, please review if these extensions for Oracle is enough.

**Comment by filippobottega** [comment]
@sdanyliv , I need to add the materialize hint. Please refer to [MATERIALIZE Hint](http://www.oracle-wiki.net/startsqlhintmaterialize) or [The undocumented Oracle SQL materialize Hint](http://www.dba-oracle.com/t_materialize_sql_hint.htm)

**Comment by sdanyliv** [comment]
Proposed design supports that
```cs
query = query.OracleSelectHint(q => "MATERIALIZE");
```

**Comment by filippobottega** [comment]
Perfect! Thank you!

**Comment by igor-tkachev** [comment]
We can add three generic methods to support different types of hints

```cs
var query =
(
   from t1 in table1.TableHint("...")
   join t2 in table2.JoinHint("...") on t1.Id equals t2.Id
   select new { t1, t2 }
)
.QueryHint("...")
```

Also, we can add specific API for each provider.

**Comment by jogibear9988** [comment]
I think it would be usefull if we can specify that a Hint should only be used for a specific provider.

**Comment by artelk** [comment]
> ```cs
> query = query.StatementHint(q => $"LEADING({q.t1} {q.t2}) USE_NL({q.t1}) INDEX({q.t1} emp_emp_id_pk) USE_MERGE({q.t1}) FULL({q.t1})")
> ```

This requires the whole objects to be selected.
```cs
var query = 
   from t1 in table1
   join t2 in table2 on t1.Id equals t2.Id
   select t1.F1 + t2.F2
```
It won't be possible to specify the hints. Of course there is a workaround: create the query with new {t1, t2}, add the hints and then create another query `var q2 = q1.Select(q => q.t1.F1 + q.t2.F2)`.

Would it be possible to implement something like this:
```cs
var query = 
   from t1 in table1
   join t2 in table2 on t1.Id equals t2.Id
   select (t1.F1 + t2.F2).AddHints(Hints.Leading(t1, t2), Hints.UseNl(t1), ...);
```

**Comment by iColdoCatalin** [comment]
Has this feature been added?
I need to add WITH NOLOCK after each table on some queries and cannot figure out how to do it.
I'm on v3.6 of linq2b an tryied out as mentioned in Query Hint Extensions #3297 the sintax
`var q =
(
	from c in db.Child
	join p in db.Parent on c.ParentID equals p.ParentID
	select p
)
.QueryHint(Hints.Option.HashJoin)`
but with no results since QueryHint is not found.

**Comment by sdanyliv** [comment]
`Table.With(“NOLOCK”)` already there, other hints not merged yet. Currently you can apply hint only to Table, not query.

**Comment by iColdoCatalin** [comment]
Thanks for info @sdanyliv

**Comment by MaceWindu** [comment]
Implemented for next v4 preview #3297
